### DISCUSSION TITLE
Memory demands for larger scale reactive transport simulations

### DISCUSSION THREAD

**Discussion by j-engelmann** [reaktoro/reaktoro#391] [Q&A]
### Description

Hello!

In my ongoing work on coupling Reaktoro and OpenFOAM to develop reactive transport models, i noticed that the memory used by the executables i create increases slowly over time (not necessarily surprising), and i believe that Reaktoro (or the way i use it) is responsible. My question is if I can control or limit this somehow, since it seems to be growing infinetly according to the number of equilibrium calculations i do. I will probably need to be able to comfortably run millions of equilibrations without clogging my memory, so small things will add up.

My setup is roughly like this, following the [suggestions in the Tutorial:](https://reaktoro.org/tutorials/equilibrium/equilibrium-with-given-component-amounts.html)

I have 1 `EquilibriumSolver` instance and 1 `ChemicalState`, and pass pressure, temperature and a chemical input vector to it through `EquilibriumConditions`. I'm using the Mines TDB through Thermofun, if that matters.

From what I can tell, the `EquilibriumSolver` and the `ThermoEngine` in it are increasing in size at each step. The screenshot below shows the "Call Trees" from XCode/Instruments/Leaks, after roughly 120 000 calculations in a relatively simple reactive transport case. The amount of memory is still quite managable there, but it grows into 10s of GB if i let it run long enough. This would also scale up massively for more complex cases, I assume.

![Screenshot 2024-07-18 at 15 04 10](https://github.com/user-attachments/assets/f6df66c0-c6ad-4ff5-a6cb-18011ded2b40)

I could imagine that there are some things worth storing to make things faster down the line (like thermodynamic properties of phases at all unique P/T conditions), is that what's happening here? If so, what could i do to avoid it going infinite?

If you have any ideas or suggestions how to approach this, let me know!

### What is your operating system?

macOS

**Comment by allanleal** [comment]
Hi @j-engelmann - I was on vacation. Have you found a solution for this in the meantime? Are you using Python or C++? I'm familiar with such problems when using Python, when `ChemicalState` objects are created each iteration and the garbage collector does not clean them over time.

I'd need a simple example that reproduces your issue so that I can debug and identify the cause.

**Reply by j-engelmann** [reply]
Hi @allanleal, thanks for the reply!

I mainly use C++ (because the transport part is also written in it), though i often use the python module to quickly check how something works. I only have one `ChemicalState` object, so i don't think that is a problem. 

I'll post the the simplest example i could come up with below. Here, i just initialize the required bits (using `EquilibriumConditions.setInitialComponentAmounts` to set the chemical composition (1kg of H2O and some Ca/S) and let it run through a nested loop of "time" and an arbitrary 1D-domain. Importantly, pressure is varied through time and space, and that somehow leads to the increased RAM demand. I assume it would be the same if I changed temperature or the bIn vector, but changing one is enough to show what i mean.

Let me know if you need anything else to look into this and thank you for your time!

Jasper


```c++
// simple program to demonstrate RAM explosions

#include <Reaktoro/Reaktoro.hpp>
using namespace Reaktoro;

int main()
{
    // initialize stuff
    ThermoFunDatabase db("mines16");
    AqueousPhase solution("H2O@ H+ H2@ O2@ OH- Ca(SO4)@ Ca+2 CaOH+ H2S@ HS- HSO4- SO4-2");
    MineralPhase anhydrite("Anhydrite");
    ChemicalSystem system(db,solution,anhydrite);
    ChemicalState state(system);
    EquilibriumConditions conditions(system);
    EquilibriumSolver solver(system);
    ChemicalProps props(state);
    
    // initialize chemical array
    ArrayXd bIn(system.elements().size()+1); // vector to hold element and charge amounts
    bIn[0] = 111.1; // H
    bIn[1] = 55.55; // O
    bIn[2] = 0.1;   // S
    bIn[3] = 0.2;   // Ca
    
    // set chemistry
    conditions.setInitialComponentAmounts(bIn);
    
    // set p, T
    conditions.temperature(323, "kelvin");
    conditions.pressure(1e5, "pascal");
    
    // number of timesteps and cells, just to simulate what happens when equilibrium is recalculated in a transport simulation
    const int nt = 1000;
    const int nx = 100000;
    
    // "time loop"
    for(int it = 0; it < nt; it++){
        // "loop over domain"
        for(int ix = 0; ix < nx; ix++){
            // change pressure (or temperature/chemistry, just change something)
            conditions.pressure(1e5+it+ix, "pascal");
            // equilibrate
            auto result = solver.solve(state,conditions);
            if(result.failed()){
                std::cout << "equilibrium not achieved." << std::endl;
            }
        }
        std::cout << it+1 << "/" << nt << " pseudo timesteps done." << std::endl;
    }
    return 0;
}

```

**Reply by allanleal** [reply]
Hi Jasper, thanks for providing this reproducible example. I'm now back in office and could check your first message more carefully. The call tree you provided is very helpful. It shows that the issue lies most likely with ThermoFun. I'll ping @gdmiron here. 

@gdmiron I think this issue could be related to maybe some memoization optimization, in which you save inputs and outputs of a function to avoid repeated calculation when same inputs are provided? In Reaktoro, this is done by saving only the very last inputs/outputs -- not all of them (i.e., history). I think you once told me that you save all inputs and corresponding outputs, but could be wrong.

**Reply by j-engelmann** [reply]
Hi @allanleal,

thanks for looking into it! That sounds like a reasonable explaination. My idea of a temporary fix would be to simply reset the whole `EquilibriumSolver` instance every now and then, say after a certain number of calculations. I tried implementing this by creating a std::shared_ptr to the solver instance and using the .reset() function of the shared pointer. That doesn't seem to work though, which may be related to my limited C++ knowledge.

In the example above, all i change is the line creating the solver to:

```c++
std::shared_ptr<EquilibriumSolver> solver = std::make_shared<EquilibriumSolver>(system);
``` 

and then do this within the outer for loop:
```c++
solver.reset();
solver = std::make_shared<EquilibriumSolver>(system);
``` 

Doing this leads to the same pattern of ever increasing memory usage, so its not much help. Can you see a way to achieve this little workaround?

**Reply by allanleal** [reply]
Hi @j-engelmann - all ThermoFun functions/objects would have to be recreated, and this happens deep during the creation of the `ChemicalSystem` object. You'd need to recreate the `ChemicalSystem`. If this does not work, I think you'd need to recreate the `ThermoFunDatabase` object and then the `ChemicalSystem` object.

@gdmiron - any way to control the memoization so that the issue above can be resolved?

**Reply by j-engelmann** [reply]
Thanks for the ideas! Sad to report that even recreating the ChemicalSystem and Database objects doesn't really change anything. For now, i can make it work by just restarting my entire program every now and then. That's not ideal, but it has helped me identify some other issues so i guess its time well spent. 

I've created an Issue on the [ThermoFun github](https://github.com/thermohub/thermofun/issues/72) aswell, so we can continue the discussion there if neeeded.

**Reply by allanleal** [reply]
Thanks @j-engelmann for creating the issue on the ThermoFun project. I'll follow the discussion there.

**Reply by j-engelmann** [reply]
As mentioned in the ThermoFun git, i have come up with a working solution myself. Either by turning of memoization in ThermoFun, or with a new template that caps cache size. If anyone else runs into this issue, i'll be happy to share. 

We can close this issue here, thanks for all the input @allanleal!

**Reply by allanleal** [reply]
@j-engelmann Thanks a lot for the update and investigating this. Time permitting, I'll try to contribute to a PR to ThermoFun to let users disable this at runtime.
